packages/adapter-components/src/references/referenced_instance_name.ts

```diff
@@ -0,0 +1,231 @@
+    ? fileNameParts.join('_')
+    : undefined))
+  return [
+    instance.elemID.adapter,
+    RECORDS_PATH,
+    pathNaclCase(typeName),
+    fileName ? pathNaclCase(naclCase(fileName)) : pathNaclCase(naclName),
+  ]
+}
+
+const getNeighborsByIdFields = (
+  instanceToIdFields: InstanceIdFields,
+  allInstancesToIdFields: InstanceIdFields[],
+): InstanceIdFields[] => {
+  const neighbors: InstanceIdFields[] = []
+  instanceToIdFields.idFields.forEach(
+    idField => {
+      if (isReferencedIdField(idField)) {
+        const fieldValue = _.get(instanceToIdFields.instance.value, idField.substr(1))
+        if (isReferenceExpression(fieldValue) && isInstanceElement(fieldValue.value)) {
+          const neighbor = fieldValue.value
+          const neighborIdFields = allInstancesToIdFields
+            .find(instanceToIdField => instanceToIdField.instance === neighbor)?.idFields
+          if (neighborIdFields !== undefined) {
+            neighbors.push({ instance: neighbor, idFields: neighborIdFields })
+          }
+        }
+      }
+    }
+  )
+  return neighbors
+}
+
+const dfs = (
+  currentInstanceIdFields: InstanceIdFields,
+  allInstancesIdFields: InstanceIdFields[],
+  sortedInstances: InstanceIdFields[],
+  visitedInstances: InstanceIdFields[],
+  stack: InstanceIdFields[],
+): void => {
+  visitedInstances.push(currentInstanceIdFields)
+  stack.push(currentInstanceIdFields)
+  const neighbors = getNeighborsByIdFields(currentInstanceIdFields, allInstancesIdFields)
+  neighbors.forEach(
+    neighbor => {
+      if (visitedInstances
+        .some(visited => (visited.instance.elemID).isEqual(neighbor.instance.elemID))
+        && stack
+          .some(instack => (instack.instance.elemID).isEqual(neighbor.instance.elemID))) {
+        log.warn('A cycle of references was detected in instances idFields')
+        throw new Error('Instances names cannot be created when there is a cycle of references in IdFields')
+      } else if (!visitedInstances
+        .some(visited => (visited.instance.elemID).isEqual(neighbor.instance.elemID))
+      ) {
+        dfs(neighbor, allInstancesIdFields, sortedInstances, visitedInstances, stack)
+      }
+    }
+  )
+  stack.pop()
+  sortedInstances.push(currentInstanceIdFields)
+}
+
+export const sortInstancesByReference = (
+  instancesToIdFields: InstanceIdFields[]
+): InstanceIdFields[] => {
+  const sortedInstances: InstanceIdFields[] = []
+  const visitedInstances: InstanceIdFields[] = []
+  const stack: InstanceIdFields[] = []
+  instancesToIdFields.forEach(
+    refInstanceToIdFields => {
+      if (!visitedInstances
+        .some(visited => (visited.instance.elemID).isEqual(refInstanceToIdFields.instance.elemID))
+      ) {
+        dfs(refInstanceToIdFields, instancesToIdFields, sortedInstances, visitedInstances, stack)
+      }
+    }
+  )
+  return sortedInstances
+}
+
+/*
+ * Change instances name for referenced instances names
```

Comment:
thanks for adding documentation :) I think this may be worth a slightly more elaborate explanation - maybe something like
```suggestion
 * Utility function that finds instance elements whose id relies on the ids of other instances, and replaces them with updated instances with the correct id and file path.
```
also - I assume this function is written this way because it is the main "filter" function - I think the "wrapping" part might fit better under the `filters` folder, wdyt?